{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b07ef072-7295-46a7-b0a1-82d243efba0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import gzip\n",
    "import wget\n",
    "import requests"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "df905fed-cde6-4516-b41b-60ece42f8a34",
   "metadata": {},
   "source": [
    "Using only Python3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "19374324-0681-4c81-8238-07461c671c59",
   "metadata": {},
   "outputs": [],
   "source": [
    "# https://www.ensembl.org/Homo_sapiens/Info/Index\n",
    "# http://ftp.ensembl.org/pub/release-104/gtf/homo_sapiens/\n",
    "# chr.gtf.gz without GL and GI unlocalized regions\n",
    "\n",
    "#os.makedirs('data', exist_ok=True)\n",
    "#url = 'http://ftp.ensembl.org/pub/release-104/gtf/homo_sapiens/Homo_sapiens.GRCh38.104.chr.gtf.gz'\n",
    "#r = requests.get(url, allow_redirects=True)\n",
    "filename = \"./data/Homo_sapiens.GRCh38.104.chr.gtf.gz\"\n",
    "#open(filename, \"wb\").write(r.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2b4520f6-9c4b-4c9c-adfd-0b9b518b1142",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[b'#!genome-build GRCh38.p13\\n', b'#!genome-version GRCh38\\n', b'#!genome-date 2013-12\\n', b'#!genome-build-accession GCA_000001405.28\\n', b'#!genebuild-last-updated 2021-03\\n', b'1\\tensembl_havana\\tgene\\t685679\\t686673\\t.\\t-\\t.\\tgene_id \"ENSG00000284662\"; gene_version \"1\"; gene_name \"OR4F16\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\";\\n', b'1\\tensembl_havana\\ttranscript\\t685679\\t686673\\t.\\t-\\t.\\tgene_id \"ENSG00000284662\"; gene_version \"1\"; transcript_id \"ENST00000332831\"; transcript_version \"4\"; gene_name \"OR4F16\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\"; transcript_name \"OR4F16-201\"; transcript_source \"ensembl_havana\"; transcript_biotype \"protein_coding\"; tag \"CCDS\"; ccds_id \"CCDS41221\"; tag \"basic\"; transcript_support_level \"NA (assigned to previous version 3)\";\\n', b'1\\tensembl_havana\\texon\\t685679\\t686673\\t.\\t-\\t.\\tgene_id \"ENSG00000284662\"; gene_version \"1\"; transcript_id \"ENST00000332831\"; transcript_version \"4\"; exon_number \"1\"; gene_name \"OR4F16\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\"; transcript_name \"OR4F16-201\"; transcript_source \"ensembl_havana\"; transcript_biotype \"protein_coding\"; tag \"CCDS\"; ccds_id \"CCDS41221\"; exon_id \"ENSE00002324228\"; exon_version \"3\"; tag \"basic\"; transcript_support_level \"NA (assigned to previous version 3)\";\\n', b'1\\tensembl_havana\\tCDS\\t685719\\t686654\\t.\\t-\\t0\\tgene_id \"ENSG00000284662\"; gene_version \"1\"; transcript_id \"ENST00000332831\"; transcript_version \"4\"; exon_number \"1\"; gene_name \"OR4F16\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\"; transcript_name \"OR4F16-201\"; transcript_source \"ensembl_havana\"; transcript_biotype \"protein_coding\"; tag \"CCDS\"; ccds_id \"CCDS41221\"; protein_id \"ENSP00000329982\"; protein_version \"2\"; tag \"basic\"; transcript_support_level \"NA (assigned to previous version 3)\";\\n', b'1\\tensembl_havana\\tstart_codon\\t686652\\t686654\\t.\\t-\\t0\\tgene_id \"ENSG00000284662\"; gene_version \"1\"; transcript_id \"ENST00000332831\"; transcript_version \"4\"; exon_number \"1\"; gene_name \"OR4F16\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\"; transcript_name \"OR4F16-201\"; transcript_source \"ensembl_havana\"; transcript_biotype \"protein_coding\"; tag \"CCDS\"; ccds_id \"CCDS41221\"; tag \"basic\"; transcript_support_level \"NA (assigned to previous version 3)\";\\n']\n"
     ]
    }
   ],
   "source": [
    "with gzip.open(filename) as f:\n",
    "    print(f.readlines()[0:10])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8321efe-d78c-49d7-948f-49e41d3f1364",
   "metadata": {},
   "source": [
    "Using bash commands through Jupyter magic comand `%%bash` and `!`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2f39c22c-5c9d-4bea-a90e-e72a097a9d8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%bash\n",
    "\n",
    "#mkdir -p data/\n",
    "\n",
    "#wget -q -P data/ http://ftp.ensembl.org/pub/release-104/gtf/homo_sapiens/Homo_sapiens.GRCh38.104.chr.gtf.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "060d62a6-8bf7-45f2-9d3b-9849886206d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "#! zcat < data/Homo_sapiens.GRCh38.104.chr.gtf.gz | head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bab02c7a-1fd7-4e42-8f7c-839f3355190c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "74fd4ee3-a853-4841-9bed-25c4eccc0388",
   "metadata": {},
   "source": [
    "The GTF (General Transfer Format) is an extension of GFF version 2 \n",
    "and used to represent transcription models. GFF (General Feature Format) \n",
    "consists of one line per feature, each containing 9 columns of data. \n",
    "\n",
    "**Fields**\n",
    "\n",
    "Fields are tab-separated. Also, all but the final field in each \n",
    "feature line must contain a value; \"empty\" columns are denoted \n",
    "with a '.'\n",
    "\n",
    "    seqname   - name of the chromosome or scaffold; chromosome names \n",
    "                without a 'chr' \n",
    "    source    - name of the program that generated this feature, or \n",
    "                the data source (database or project name)\n",
    "    feature   - feature type name. Current allowed features are\n",
    "                {gene, transcript, exon, CDS, Selenocysteine, start_codon,\n",
    "                stop_codon and UTR}\n",
    "    start     - start position of the feature, with sequence numbering \n",
    "                starting at 1.\n",
    "    end       - end position of the feature, with sequence numbering \n",
    "                starting at 1.\n",
    "    score     - a floating point value indiciating the score of a feature\n",
    "    strand    - defined as + (forward) or - (reverse).\n",
    "    frame     - one of '0', '1' or '2'. Frame indicates the number of base pairs\n",
    "                before you encounter a full codon. '0' indicates the feature \n",
    "                begins with a whole codon. '1' indicates there is an extra\n",
    "                base (the 3rd base of the prior codon) at the start of this feature.\n",
    "                '2' indicates there are two extra bases (2nd and 3rd base of the \n",
    "                prior exon) before the first codon. All values are given with\n",
    "                relation to the 5' end.\n",
    "    attribute - a semicolon-separated list of tag-value pairs (separated by a space), \n",
    "                providing additional information about each feature. A key can be\n",
    "                repeated multiple times.\n",
    "\n",
    "**Attributes**\n",
    "\n",
    "The following attributes are available. All attributes are semi-colon\n",
    "separated pairs of keys and values.\n",
    "\n",
    "    - gene_id: The stable identifier for the gene\n",
    "    - gene_version: The stable identifier version for the gene\n",
    "    - gene_name: The official symbol of this gene\n",
    "    - gene_source: The annotation source for this gene\n",
    "    - gene_biotype: The biotype of this gene\n",
    "    - transcript_id: The stable identifier for this transcript\n",
    "    - transcript_version: The stable identifier version for this transcript\n",
    "    - transcript_name: The symbold for this transcript derived from the gene name\n",
    "    - transcript_source: The annotation source for this transcript\n",
    "    - transcript_biotype: The biotype for this transcript\n",
    "    - exon_id: The stable identifier for this exon\n",
    "    - exon_version: The stable identifier version for this exon\n",
    "    - exon_number: Position of this exon in the transcript\n",
    "    - ccds_id: CCDS identifier linked to this transcript\n",
    "    - protein_id: Stable identifier for this transcript's protein\n",
    "    - protein_version: Stable identifier version for this transcript's protein\n",
    "    - tag: A collection of additional key value tags\n",
    "    - transcript_support_level: Ranking to assess how well a transcript is supported (from 1 to 5)\n",
    "\n",
    "**Tags**\n",
    "\n",
    "Tags are additional flags used to indicate attibutes of the transcript.\n",
    "\n",
    "    - CCDS: Flags this transcript as one linked to a CCDS record\n",
    "    - seleno: Flags this transcript has a Selenocysteine edit.\n",
    "    - cds_end_NF: the coding region end could not be confirmed\n",
    "    - cds_start_NF: the coding region start could not be confirmed\n",
    "    - mRNA_end_NF: the mRNA end could not be confirmed\n",
    "    - mRNA_start_NF: the mRNA start could not be confirmed.\n",
    "    - basic: the transcript is part of the gencode basic geneset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d8eb40b2-e02b-4944-8fc4-4d4f28fe6229",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv('data/Homo_sapiens.GRCh38.104.chr.gtf.gz', \n",
    "            sep='\\t', \n",
    "            comment='#',\n",
    "            header=None,\n",
    "            dtype=str,\n",
    "            names=['seqname', 'source', 'feature', 'start', \n",
    "                   'end', 'score', 'strand', 'frame', \n",
    "                   'attribute'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ba7ee3ca-33fc-46e1-8f73-c117b8b40a07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(3145418, 9)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "90eddb3b-c28c-4c4d-8cb3-2c31dbdbcb0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>685679</td>\n",
       "      <td>686673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>685679</td>\n",
       "      <td>686673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>685679</td>\n",
       "      <td>686673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>685719</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>686652</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname          source      feature   start     end score strand frame  \\\n",
       "0       1  ensembl_havana         gene  685679  686673     .      -     .   \n",
       "1       1  ensembl_havana   transcript  685679  686673     .      -     .   \n",
       "2       1  ensembl_havana         exon  685679  686673     .      -     .   \n",
       "3       1  ensembl_havana          CDS  685719  686654     .      -     0   \n",
       "4       1  ensembl_havana  start_codon  686652  686654     .      -     0   \n",
       "\n",
       "                                           attribute  \n",
       "0  gene_id \"ENSG00000284662\"; gene_version \"1\"; g...  \n",
       "1  gene_id \"ENSG00000284662\"; gene_version \"1\"; t...  \n",
       "2  gene_id \"ENSG00000284662\"; gene_version \"1\"; t...  \n",
       "3  gene_id \"ENSG00000284662\"; gene_version \"1\"; t...  \n",
       "4  gene_id \"ENSG00000284662\"; gene_version \"1\"; t...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "e756ecf0-7fc0-4d08-9eee-e2715306e6a8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>3145413</th>\n",
       "      <td>MT</td>\n",
       "      <td>insdc</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210195\"; gene_version \"2\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3145414</th>\n",
       "      <td>MT</td>\n",
       "      <td>insdc</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210195\"; gene_version \"2\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3145415</th>\n",
       "      <td>MT</td>\n",
       "      <td>insdc</td>\n",
       "      <td>gene</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210196\"; gene_version \"2\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3145416</th>\n",
       "      <td>MT</td>\n",
       "      <td>insdc</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210196\"; gene_version \"2\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3145417</th>\n",
       "      <td>MT</td>\n",
       "      <td>insdc</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210196\"; gene_version \"2\"; t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqname source     feature  start    end score strand frame  \\\n",
       "3145413      MT  insdc  transcript  15888  15953     .      +     .   \n",
       "3145414      MT  insdc        exon  15888  15953     .      +     .   \n",
       "3145415      MT  insdc        gene  15956  16023     .      -     .   \n",
       "3145416      MT  insdc  transcript  15956  16023     .      -     .   \n",
       "3145417      MT  insdc        exon  15956  16023     .      -     .   \n",
       "\n",
       "                                                 attribute  \n",
       "3145413  gene_id \"ENSG00000210195\"; gene_version \"2\"; t...  \n",
       "3145414  gene_id \"ENSG00000210195\"; gene_version \"2\"; t...  \n",
       "3145415  gene_id \"ENSG00000210196\"; gene_version \"2\"; g...  \n",
       "3145416  gene_id \"ENSG00000210196\"; gene_version \"2\"; t...  \n",
       "3145417  gene_id \"ENSG00000210196\"; gene_version \"2\"; t...  "
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ccbf71ad-fca2-46d1-803b-4e0cc84c1980",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "seqname      object\n",
       "source       object\n",
       "feature      object\n",
       "start        object\n",
       "end          object\n",
       "score        object\n",
       "strand       object\n",
       "frame        object\n",
       "attribute    object\n",
       "dtype: object"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.dtypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b05549f6-0774-4828-b827-846a5b9451c6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "seqname      0\n",
       "source       0\n",
       "feature      0\n",
       "start        0\n",
       "end          0\n",
       "score        0\n",
       "strand       0\n",
       "frame        0\n",
       "attribute    0\n",
       "dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.isna().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "b733ea51-0094-4864-8d72-0e84ffea17a1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.isna().sum().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "ab631d11-6f01-4f7a-99ac-473cc787ed01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['seqname', 'source', 'feature', 'start', 'end', 'score', 'strand',\n",
       "       'frame', 'attribute'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "79d14763-8005-4aef-8372-73f26ecf607f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['1', '2', '3', '4', '5', '6', '7', 'X', '8', '9', '11', '10', '12',\n",
       "       '13', '14', '15', '16', '17', '18', '20', '19', 'Y', '22', '21',\n",
       "       'MT'], dtype=object)"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['seqname'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "04e14069-3bd0-4fc9-967d-271ecf5fe166",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "25"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df['seqname'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "df5180b9-8781-4f0d-babf-e79e85943e0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "seqname 25\n",
      "source 7\n",
      "feature 9\n",
      "start 639387\n",
      "end 639576\n",
      "score 1\n",
      "strand 2\n",
      "frame 4\n",
      "attribute 2794860\n"
     ]
    }
   ],
   "source": [
    "for x in df.columns:\n",
    "    print(x, len(df[x].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "463452cd-55ae-4076-ad03-c04b2fcb6599",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['seqname', 25],\n",
       " ['source', 7],\n",
       " ['feature', 9],\n",
       " ['start', 639387],\n",
       " ['end', 639576],\n",
       " ['score', 1],\n",
       " ['strand', 2],\n",
       " ['frame', 4],\n",
       " ['attribute', 2794860]]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[[x, len(df[x].unique())] for x in df.columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "03f249a5-6d1b-4bf1-9ed0-8741ebf7e1b4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "seqname ['1' '2' '3' '4' '5' '6' '7' 'X' '8' '9' '11' '10' '12' '13' '14' '15'\n",
      " '16' '17' '18' '20' '19' 'Y' '22' '21' 'MT']\n",
      "source ['ensembl_havana' 'havana' 'ensembl' 'havana_tagene'\n",
      " 'ensembl_havana_tagene' 'mirbase' 'insdc']\n",
      "feature ['gene' 'transcript' 'exon' 'CDS' 'start_codon' 'stop_codon'\n",
      " 'five_prime_utr' 'three_prime_utr' 'Selenocysteine']\n",
      "score ['.']\n",
      "strand ['-' '+']\n",
      "frame ['.' '0' '2' '1']\n"
     ]
    }
   ],
   "source": [
    "for x in df.columns:\n",
    "    if len(df[x].unique()) < 30:\n",
    "        print(x, df[x].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5593c115-2f54-4904-8095-3e67cdd28000",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['seqname',\n",
       "  array(['1', '2', '3', '4', '5', '6', '7', 'X', '8', '9', '11', '10', '12',\n",
       "         '13', '14', '15', '16', '17', '18', '20', '19', 'Y', '22', '21',\n",
       "         'MT'], dtype=object)],\n",
       " ['source',\n",
       "  array(['ensembl_havana', 'havana', 'ensembl', 'havana_tagene',\n",
       "         'ensembl_havana_tagene', 'mirbase', 'insdc'], dtype=object)],\n",
       " ['feature',\n",
       "  array(['gene', 'transcript', 'exon', 'CDS', 'start_codon', 'stop_codon',\n",
       "         'five_prime_utr', 'three_prime_utr', 'Selenocysteine'],\n",
       "        dtype=object)],\n",
       " ['score', array(['.'], dtype=object)],\n",
       " ['strand', array(['-', '+'], dtype=object)],\n",
       " ['frame', array(['.', '0', '2', '1'], dtype=object)]]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[[x, df[x].unique()] for x in df.columns if len(df[x].unique()) < 30 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "6be7d2c7-0e23-4011-823a-adaa4a4dbfcb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pandas.core.groupby.generic.DataFrameGroupBy object at 0x7fae2ec3a310>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby('feature')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "0bf994ef-dbcd-4d6d-900f-d9a5685f831e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>feature</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>CDS</th>\n",
       "      <td>824576</td>\n",
       "      <td>824576</td>\n",
       "      <td>824576</td>\n",
       "      <td>824576</td>\n",
       "      <td>824576</td>\n",
       "      <td>824576</td>\n",
       "      <td>824576</td>\n",
       "      <td>824576</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Selenocysteine</th>\n",
       "      <td>119</td>\n",
       "      <td>119</td>\n",
       "      <td>119</td>\n",
       "      <td>119</td>\n",
       "      <td>119</td>\n",
       "      <td>119</td>\n",
       "      <td>119</td>\n",
       "      <td>119</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>exon</th>\n",
       "      <td>1498154</td>\n",
       "      <td>1498154</td>\n",
       "      <td>1498154</td>\n",
       "      <td>1498154</td>\n",
       "      <td>1498154</td>\n",
       "      <td>1498154</td>\n",
       "      <td>1498154</td>\n",
       "      <td>1498154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>five_prime_utr</th>\n",
       "      <td>163233</td>\n",
       "      <td>163233</td>\n",
       "      <td>163233</td>\n",
       "      <td>163233</td>\n",
       "      <td>163233</td>\n",
       "      <td>163233</td>\n",
       "      <td>163233</td>\n",
       "      <td>163233</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene</th>\n",
       "      <td>60605</td>\n",
       "      <td>60605</td>\n",
       "      <td>60605</td>\n",
       "      <td>60605</td>\n",
       "      <td>60605</td>\n",
       "      <td>60605</td>\n",
       "      <td>60605</td>\n",
       "      <td>60605</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>start_codon</th>\n",
       "      <td>92911</td>\n",
       "      <td>92911</td>\n",
       "      <td>92911</td>\n",
       "      <td>92911</td>\n",
       "      <td>92911</td>\n",
       "      <td>92911</td>\n",
       "      <td>92911</td>\n",
       "      <td>92911</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>stop_codon</th>\n",
       "      <td>86172</td>\n",
       "      <td>86172</td>\n",
       "      <td>86172</td>\n",
       "      <td>86172</td>\n",
       "      <td>86172</td>\n",
       "      <td>86172</td>\n",
       "      <td>86172</td>\n",
       "      <td>86172</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>three_prime_utr</th>\n",
       "      <td>182795</td>\n",
       "      <td>182795</td>\n",
       "      <td>182795</td>\n",
       "      <td>182795</td>\n",
       "      <td>182795</td>\n",
       "      <td>182795</td>\n",
       "      <td>182795</td>\n",
       "      <td>182795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>transcript</th>\n",
       "      <td>236853</td>\n",
       "      <td>236853</td>\n",
       "      <td>236853</td>\n",
       "      <td>236853</td>\n",
       "      <td>236853</td>\n",
       "      <td>236853</td>\n",
       "      <td>236853</td>\n",
       "      <td>236853</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 seqname   source    start      end    score   strand  \\\n",
       "feature                                                                 \n",
       "CDS               824576   824576   824576   824576   824576   824576   \n",
       "Selenocysteine       119      119      119      119      119      119   \n",
       "exon             1498154  1498154  1498154  1498154  1498154  1498154   \n",
       "five_prime_utr    163233   163233   163233   163233   163233   163233   \n",
       "gene               60605    60605    60605    60605    60605    60605   \n",
       "start_codon        92911    92911    92911    92911    92911    92911   \n",
       "stop_codon         86172    86172    86172    86172    86172    86172   \n",
       "three_prime_utr   182795   182795   182795   182795   182795   182795   \n",
       "transcript        236853   236853   236853   236853   236853   236853   \n",
       "\n",
       "                   frame  attribute  \n",
       "feature                              \n",
       "CDS               824576     824576  \n",
       "Selenocysteine       119        119  \n",
       "exon             1498154    1498154  \n",
       "five_prime_utr    163233     163233  \n",
       "gene               60605      60605  \n",
       "start_codon        92911      92911  \n",
       "stop_codon         86172      86172  \n",
       "three_prime_utr   182795     182795  \n",
       "transcript        236853     236853  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby('feature').count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "eb54252f-b4a6-4fe0-ac52-2dcde211b4ed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "feature\n",
       "CDS                 824576\n",
       "Selenocysteine         119\n",
       "exon               1498154\n",
       "five_prime_utr      163233\n",
       "gene                 60605\n",
       "start_codon          92911\n",
       "stop_codon           86172\n",
       "three_prime_utr     182795\n",
       "transcript          236853\n",
       "Name: feature, dtype: int64"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby('feature')['feature'].count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "844cf134-58f0-483b-8906-f545dc287c13",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[seqname\n",
       " 1     286554\n",
       " 10    120959\n",
       " 11    189276\n",
       " 12    179072\n",
       " 13     50861\n",
       " 14    114641\n",
       " 15    114177\n",
       " 16    142494\n",
       " 17    187948\n",
       " 18     55262\n",
       " 19    179208\n",
       " 2     238910\n",
       " 20     71221\n",
       " 21     35568\n",
       " 22     65887\n",
       " 3     200212\n",
       " 4     130868\n",
       " 5     141226\n",
       " 6     149532\n",
       " 7     149512\n",
       " 8     116505\n",
       " 9     114334\n",
       " MT       144\n",
       " X     102923\n",
       " Y       8124\n",
       " Name: seqname, dtype: int64,\n",
       " source\n",
       " ensembl                   186372\n",
       " ensembl_havana            707584\n",
       " ensembl_havana_tagene        152\n",
       " havana                   2115992\n",
       " havana_tagene             129543\n",
       " insdc                        144\n",
       " mirbase                     5631\n",
       " Name: source, dtype: int64,\n",
       " feature\n",
       " CDS                 824576\n",
       " Selenocysteine         119\n",
       " exon               1498154\n",
       " five_prime_utr      163233\n",
       " gene                 60605\n",
       " start_codon          92911\n",
       " stop_codon           86172\n",
       " three_prime_utr     182795\n",
       " transcript          236853\n",
       " Name: feature, dtype: int64,\n",
       " score\n",
       " .    3145418\n",
       " Name: score, dtype: int64,\n",
       " strand\n",
       " +    1609856\n",
       " -    1535562\n",
       " Name: strand, dtype: int64,\n",
       " frame\n",
       " .    2141759\n",
       " 0     606969\n",
       " 1     162010\n",
       " 2     234680\n",
       " Name: frame, dtype: int64]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[df.groupby(x)[x].count() for x in df.columns if len(df[x].unique()) < 30]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d71ea6a-5f64-4a4f-8f80-7069c82e064d",
   "metadata": {},
   "source": [
    "Crear un archivo bed para genes y otro para transcritos"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "59fd2c37-cb62-4566-9e64-a03c46e6f30d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>685679</td>\n",
       "      <td>686673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>685679</td>\n",
       "      <td>686673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>685679</td>\n",
       "      <td>686673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>685719</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>686652</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname          source      feature   start     end score strand frame  \\\n",
       "0       1  ensembl_havana         gene  685679  686673     .      -     .   \n",
       "1       1  ensembl_havana   transcript  685679  686673     .      -     .   \n",
       "2       1  ensembl_havana         exon  685679  686673     .      -     .   \n",
       "3       1  ensembl_havana          CDS  685719  686654     .      -     0   \n",
       "4       1  ensembl_havana  start_codon  686652  686654     .      -     0   \n",
       "\n",
       "                                           attribute  \n",
       "0  gene_id \"ENSG00000284662\"; gene_version \"1\"; g...  \n",
       "1  gene_id \"ENSG00000284662\"; gene_version \"1\"; t...  \n",
       "2  gene_id \"ENSG00000284662\"; gene_version \"1\"; t...  \n",
       "3  gene_id \"ENSG00000284662\"; gene_version \"1\"; t...  \n",
       "4  gene_id \"ENSG00000284662\"; gene_version \"1\"; t...  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "7d3de498-3744-4cc3-ab8b-6c0e1518966e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    gene_id \"ENSG00000284662\"; gene_version \"1\"; g...\n",
       "1    gene_id \"ENSG00000284662\"; gene_version \"1\"; t...\n",
       "2    gene_id \"ENSG00000284662\"; gene_version \"1\"; t...\n",
       "3    gene_id \"ENSG00000284662\"; gene_version \"1\"; t...\n",
       "4    gene_id \"ENSG00000284662\"; gene_version \"1\"; t...\n",
       "5    gene_id \"ENSG00000284662\"; gene_version \"1\"; t...\n",
       "6    gene_id \"ENSG00000284662\"; gene_version \"1\"; t...\n",
       "7    gene_id \"ENSG00000284662\"; gene_version \"1\"; t...\n",
       "8    gene_id \"ENSG00000186827\"; gene_version \"11\"; ...\n",
       "9    gene_id \"ENSG00000186827\"; gene_version \"11\"; ...\n",
       "Name: attribute, dtype: object"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['attribute'][0:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "5b9b1f0e-c2b5-4087-9fff-430c11485cef",
   "metadata": {},
   "outputs": [],
   "source": [
    "# more examples of slicing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "4a767a6f-21fc-49e7-9484-617e1d2a170e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['gene_id \"ENSG00000284662\"; gene_version \"1\"; gene_name \"OR4F16\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\";',\n",
       " 'gene_id \"ENSG00000284662\"; gene_version \"1\"; transcript_id \"ENST00000332831\"; transcript_version \"4\"; gene_name \"OR4F16\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\"; transcript_name \"OR4F16-201\"; transcript_source \"ensembl_havana\"; transcript_biotype \"protein_coding\"; tag \"CCDS\"; ccds_id \"CCDS41221\"; tag \"basic\"; transcript_support_level \"NA (assigned to previous version 3)\";',\n",
       " 'gene_id \"ENSG00000284662\"; gene_version \"1\"; transcript_id \"ENST00000332831\"; transcript_version \"4\"; exon_number \"1\"; gene_name \"OR4F16\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\"; transcript_name \"OR4F16-201\"; transcript_source \"ensembl_havana\"; transcript_biotype \"protein_coding\"; tag \"CCDS\"; ccds_id \"CCDS41221\"; exon_id \"ENSE00002324228\"; exon_version \"3\"; tag \"basic\"; transcript_support_level \"NA (assigned to previous version 3)\";',\n",
       " 'gene_id \"ENSG00000284662\"; gene_version \"1\"; transcript_id \"ENST00000332831\"; transcript_version \"4\"; exon_number \"1\"; gene_name \"OR4F16\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\"; transcript_name \"OR4F16-201\"; transcript_source \"ensembl_havana\"; transcript_biotype \"protein_coding\"; tag \"CCDS\"; ccds_id \"CCDS41221\"; protein_id \"ENSP00000329982\"; protein_version \"2\"; tag \"basic\"; transcript_support_level \"NA (assigned to previous version 3)\";',\n",
       " 'gene_id \"ENSG00000284662\"; gene_version \"1\"; transcript_id \"ENST00000332831\"; transcript_version \"4\"; exon_number \"1\"; gene_name \"OR4F16\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\"; transcript_name \"OR4F16-201\"; transcript_source \"ensembl_havana\"; transcript_biotype \"protein_coding\"; tag \"CCDS\"; ccds_id \"CCDS41221\"; tag \"basic\"; transcript_support_level \"NA (assigned to previous version 3)\";',\n",
       " 'gene_id \"ENSG00000284662\"; gene_version \"1\"; transcript_id \"ENST00000332831\"; transcript_version \"4\"; exon_number \"1\"; gene_name \"OR4F16\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\"; transcript_name \"OR4F16-201\"; transcript_source \"ensembl_havana\"; transcript_biotype \"protein_coding\"; tag \"CCDS\"; ccds_id \"CCDS41221\"; tag \"basic\"; transcript_support_level \"NA (assigned to previous version 3)\";',\n",
       " 'gene_id \"ENSG00000284662\"; gene_version \"1\"; transcript_id \"ENST00000332831\"; transcript_version \"4\"; gene_name \"OR4F16\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\"; transcript_name \"OR4F16-201\"; transcript_source \"ensembl_havana\"; transcript_biotype \"protein_coding\"; tag \"CCDS\"; ccds_id \"CCDS41221\"; tag \"basic\"; transcript_support_level \"NA (assigned to previous version 3)\";',\n",
       " 'gene_id \"ENSG00000284662\"; gene_version \"1\"; transcript_id \"ENST00000332831\"; transcript_version \"4\"; gene_name \"OR4F16\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\"; transcript_name \"OR4F16-201\"; transcript_source \"ensembl_havana\"; transcript_biotype \"protein_coding\"; tag \"CCDS\"; ccds_id \"CCDS41221\"; tag \"basic\"; transcript_support_level \"NA (assigned to previous version 3)\";',\n",
       " 'gene_id \"ENSG00000186827\"; gene_version \"11\"; gene_name \"TNFRSF4\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\";',\n",
       " 'gene_id \"ENSG00000186827\"; gene_version \"11\"; transcript_id \"ENST00000379236\"; transcript_version \"4\"; gene_name \"TNFRSF4\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\"; transcript_name \"TNFRSF4-201\"; transcript_source \"ensembl_havana\"; transcript_biotype \"protein_coding\"; tag \"CCDS\"; ccds_id \"CCDS11\"; tag \"basic\"; transcript_support_level \"1 (assigned to previous version 3)\";']"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[x for x in df['attribute'][0:10]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "69caf080-3c9d-4ec3-854a-b4a720baa4e6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0          False\n",
       "1          False\n",
       "2          False\n",
       "3          False\n",
       "4          False\n",
       "           ...  \n",
       "3145413    False\n",
       "3145414    False\n",
       "3145415    False\n",
       "3145416    False\n",
       "3145417    False\n",
       "Name: attribute, Length: 3145418, dtype: bool"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['attribute'] == 'ENSG00000284662'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "ea906348-71c6-4079-8d25-3e381a60899c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0           True\n",
       "1           True\n",
       "2           True\n",
       "3           True\n",
       "4           True\n",
       "           ...  \n",
       "3145413    False\n",
       "3145414    False\n",
       "3145415    False\n",
       "3145416    False\n",
       "3145417    False\n",
       "Name: attribute, Length: 3145418, dtype: bool"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['attribute'].str.contains('ENSG00000284662')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "e0ca6ee5-2aea-49d6-96e3-b7780e0a67b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>685679</td>\n",
       "      <td>686673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>685679</td>\n",
       "      <td>686673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>685679</td>\n",
       "      <td>686673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>685719</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>686652</td>\n",
       "      <td>686654</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>stop_codon</td>\n",
       "      <td>685716</td>\n",
       "      <td>685718</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>five_prime_utr</td>\n",
       "      <td>686655</td>\n",
       "      <td>686673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>three_prime_utr</td>\n",
       "      <td>685679</td>\n",
       "      <td>685715</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname          source          feature   start     end score strand frame  \\\n",
       "0       1  ensembl_havana             gene  685679  686673     .      -     .   \n",
       "1       1  ensembl_havana       transcript  685679  686673     .      -     .   \n",
       "2       1  ensembl_havana             exon  685679  686673     .      -     .   \n",
       "3       1  ensembl_havana              CDS  685719  686654     .      -     0   \n",
       "4       1  ensembl_havana      start_codon  686652  686654     .      -     0   \n",
       "5       1  ensembl_havana       stop_codon  685716  685718     .      -     0   \n",
       "6       1  ensembl_havana   five_prime_utr  686655  686673     .      -     .   \n",
       "7       1  ensembl_havana  three_prime_utr  685679  685715     .      -     .   \n",
       "\n",
       "                                           attribute  \n",
       "0  gene_id \"ENSG00000284662\"; gene_version \"1\"; g...  \n",
       "1  gene_id \"ENSG00000284662\"; gene_version \"1\"; t...  \n",
       "2  gene_id \"ENSG00000284662\"; gene_version \"1\"; t...  \n",
       "3  gene_id \"ENSG00000284662\"; gene_version \"1\"; t...  \n",
       "4  gene_id \"ENSG00000284662\"; gene_version \"1\"; t...  \n",
       "5  gene_id \"ENSG00000284662\"; gene_version \"1\"; t...  \n",
       "6  gene_id \"ENSG00000284662\"; gene_version \"1\"; t...  \n",
       "7  gene_id \"ENSG00000284662\"; gene_version \"1\"; t...  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df['attribute'].str.contains('ENSG00000284662')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "cfcf7e0f-2ca0-4e54-951f-26fc202c6521",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df.loc[df['feature'].isin(['gene', 'transcript'])].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "6c77d31a-675d-4fb4-93a5-000a6789542f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>685679</td>\n",
       "      <td>686673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>685679</td>\n",
       "      <td>686673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1211340</td>\n",
       "      <td>1214153</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186827\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1211340</td>\n",
       "      <td>1214153</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186827\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1211340</td>\n",
       "      <td>1214138</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186827\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1212019</td>\n",
       "      <td>1213498</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186827\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1203508</td>\n",
       "      <td>1206592</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186891\"; gene_version \"14\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1203508</td>\n",
       "      <td>1206571</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186891\"; gene_version \"14\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1203508</td>\n",
       "      <td>1206592</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186891\"; gene_version \"14\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1</td>\n",
       "      <td>havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1203844</td>\n",
       "      <td>1205680</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186891\"; gene_version \"14\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname          source     feature    start      end score strand frame  \\\n",
       "0       1  ensembl_havana        gene   685679   686673     .      -     .   \n",
       "1       1  ensembl_havana  transcript   685679   686673     .      -     .   \n",
       "2       1  ensembl_havana        gene  1211340  1214153     .      -     .   \n",
       "3       1  ensembl_havana  transcript  1211340  1214153     .      -     .   \n",
       "4       1          havana  transcript  1211340  1214138     .      -     .   \n",
       "5       1          havana  transcript  1212019  1213498     .      -     .   \n",
       "6       1  ensembl_havana        gene  1203508  1206592     .      -     .   \n",
       "7       1  ensembl_havana  transcript  1203508  1206571     .      -     .   \n",
       "8       1  ensembl_havana  transcript  1203508  1206592     .      -     .   \n",
       "9       1          havana  transcript  1203844  1205680     .      -     .   \n",
       "\n",
       "                                           attribute  \n",
       "0  gene_id \"ENSG00000284662\"; gene_version \"1\"; g...  \n",
       "1  gene_id \"ENSG00000284662\"; gene_version \"1\"; t...  \n",
       "2  gene_id \"ENSG00000186827\"; gene_version \"11\"; ...  \n",
       "3  gene_id \"ENSG00000186827\"; gene_version \"11\"; ...  \n",
       "4  gene_id \"ENSG00000186827\"; gene_version \"11\"; ...  \n",
       "5  gene_id \"ENSG00000186827\"; gene_version \"11\"; ...  \n",
       "6  gene_id \"ENSG00000186891\"; gene_version \"14\"; ...  \n",
       "7  gene_id \"ENSG00000186891\"; gene_version \"14\"; ...  \n",
       "8  gene_id \"ENSG00000186891\"; gene_version \"14\"; ...  \n",
       "9  gene_id \"ENSG00000186891\"; gene_version \"14\"; ...  "
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "ee58000c-b79e-4d7e-b396-39326d979b71",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['gene_id \"ENSG00000186827\"; gene_version \"11\"; gene_name \"TNFRSF4\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\";',\n",
       " 'gene_id \"ENSG00000186827\"; gene_version \"11\"; transcript_id \"ENST00000379236\"; transcript_version \"4\"; gene_name \"TNFRSF4\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\"; transcript_name \"TNFRSF4-201\"; transcript_source \"ensembl_havana\"; transcript_biotype \"protein_coding\"; tag \"CCDS\"; ccds_id \"CCDS11\"; tag \"basic\"; transcript_support_level \"1 (assigned to previous version 3)\";',\n",
       " 'gene_id \"ENSG00000186827\"; gene_version \"11\"; transcript_id \"ENST00000497869\"; transcript_version \"5\"; gene_name \"TNFRSF4\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\"; transcript_name \"TNFRSF4-203\"; transcript_source \"havana\"; transcript_biotype \"retained_intron\"; transcript_support_level \"2\";',\n",
       " 'gene_id \"ENSG00000186827\"; gene_version \"11\"; transcript_id \"ENST00000453580\"; transcript_version \"1\"; gene_name \"TNFRSF4\"; gene_source \"ensembl_havana\"; gene_biotype \"protein_coding\"; transcript_name \"TNFRSF4-202\"; transcript_source \"havana\"; transcript_biotype \"processed_transcript\"; transcript_support_level \"3\";']"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[x for x in df[df['attribute'].str.contains('ENSG00000186827')]['attribute']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "bcae73be-4254-4a6f-b5f8-b56296d5eb4c",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = df[df['source'] == 'ensembl_havana'].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "2313645a-650a-41e7-90c0-58bef57dcd06",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>685679</td>\n",
       "      <td>686673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>685679</td>\n",
       "      <td>686673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1211340</td>\n",
       "      <td>1214153</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186827\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1211340</td>\n",
       "      <td>1214153</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186827\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1203508</td>\n",
       "      <td>1206592</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186891\"; gene_version \"14\"; ...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname          source     feature    start      end score strand frame  \\\n",
       "0       1  ensembl_havana        gene   685679   686673     .      -     .   \n",
       "1       1  ensembl_havana  transcript   685679   686673     .      -     .   \n",
       "2       1  ensembl_havana        gene  1211340  1214153     .      -     .   \n",
       "3       1  ensembl_havana  transcript  1211340  1214153     .      -     .   \n",
       "4       1  ensembl_havana        gene  1203508  1206592     .      -     .   \n",
       "\n",
       "                                           attribute  \n",
       "0  gene_id \"ENSG00000284662\"; gene_version \"1\"; g...  \n",
       "1  gene_id \"ENSG00000284662\"; gene_version \"1\"; t...  \n",
       "2  gene_id \"ENSG00000186827\"; gene_version \"11\"; ...  \n",
       "3  gene_id \"ENSG00000186827\"; gene_version \"11\"; ...  \n",
       "4  gene_id \"ENSG00000186891\"; gene_version \"14\"; ...  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "759f1f38-5e7d-4e80-86e6-d4a7bc449d62",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['1', '2', '3', '4', '5', '6', '7', 'X', '8', '9', '11', '10', '12',\n",
       "       '13', '14', '15', '16', '17', '18', '20', '19', 'Y', '22', '21'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['seqname'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "5f73e686-5ac6-471d-b52e-98257c01db91",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['ensembl_havana'], dtype=object)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['source'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "e4f3f111-e52a-43f3-a9fa-fa07aa8500cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['seqname',\n",
       "  array(['1', '2', '3', '4', '5', '6', '7', 'X', '8', '9', '11', '10', '12',\n",
       "         '13', '14', '15', '16', '17', '18', '20', '19', 'Y', '22', '21'],\n",
       "        dtype=object)],\n",
       " ['source', array(['ensembl_havana'], dtype=object)],\n",
       " ['feature', array(['gene', 'transcript'], dtype=object)],\n",
       " ['score', array(['.'], dtype=object)],\n",
       " ['strand', array(['-', '+'], dtype=object)],\n",
       " ['frame', array(['.'], dtype=object)]]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[[x, df[x].unique()] for x in df.columns if len(df[x].unique()) < 30 ]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "1b71eb52-1f3c-4c11-ac27-47a807d73579",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "      <th>9</th>\n",
       "      <th>10</th>\n",
       "      <th>11</th>\n",
       "      <th>12</th>\n",
       "      <th>13</th>\n",
       "      <th>14</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000284662</td>\n",
       "      <td>1</td>\n",
       "      <td>ENST00000332831</td>\n",
       "      <td>4</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>OR4F16-201</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS41221</td>\n",
       "      <td>basic</td>\n",
       "      <td>NA (assigned to previous version 3)</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>11</td>\n",
       "      <td>ENST00000379236</td>\n",
       "      <td>4</td>\n",
       "      <td>TNFRSF4</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TNFRSF4-201</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS11</td>\n",
       "      <td>basic</td>\n",
       "      <td>1 (assigned to previous version 3)</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSG00000186891</td>\n",
       "      <td>14</td>\n",
       "      <td>ENST00000328596</td>\n",
       "      <td>10</td>\n",
       "      <td>TNFRSF18</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TNFRSF18-201</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS9</td>\n",
       "      <td>basic</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSG00000186891</td>\n",
       "      <td>14</td>\n",
       "      <td>ENST00000379268</td>\n",
       "      <td>7</td>\n",
       "      <td>TNFRSF18</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TNFRSF18-203</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS10</td>\n",
       "      <td>basic</td>\n",
       "      <td>1 (assigned to previous version 6)</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSG00000160072</td>\n",
       "      <td>20</td>\n",
       "      <td>ENST00000673477</td>\n",
       "      <td>1</td>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>ATAD3B-206</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS30</td>\n",
       "      <td>basic</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45850</th>\n",
       "      <td>ENSG00000155313</td>\n",
       "      <td>16</td>\n",
       "      <td>ENST00000351097</td>\n",
       "      <td>9</td>\n",
       "      <td>USP25</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>USP25-203</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>basic</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45851</th>\n",
       "      <td>ENSG00000155313</td>\n",
       "      <td>16</td>\n",
       "      <td>ENST00000285681</td>\n",
       "      <td>6</td>\n",
       "      <td>USP25</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>USP25-202</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS63337</td>\n",
       "      <td>basic</td>\n",
       "      <td>1</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45852</th>\n",
       "      <td>ENSG00000155313</td>\n",
       "      <td>16</td>\n",
       "      <td>ENST00000400183</td>\n",
       "      <td>7</td>\n",
       "      <td>USP25</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>USP25-204</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS63336</td>\n",
       "      <td>basic</td>\n",
       "      <td>1 (assigned to previous version 6)</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45854</th>\n",
       "      <td>ENSG00000276076</td>\n",
       "      <td>5</td>\n",
       "      <td>ENST00000619537</td>\n",
       "      <td>5</td>\n",
       "      <td>CRYAA2</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CRYAA2-201</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS82651</td>\n",
       "      <td>basic</td>\n",
       "      <td>1 (assigned to previous version 4)</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45856</th>\n",
       "      <td>ENSG00000156299</td>\n",
       "      <td>13</td>\n",
       "      <td>ENST00000286827</td>\n",
       "      <td>7</td>\n",
       "      <td>TIAM1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>TIAM1-201</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CCDS</td>\n",
       "      <td>CCDS13609</td>\n",
       "      <td>basic</td>\n",
       "      <td>5</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>26555 rows  15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    0   1                2   3         4               5   \\\n",
       "1      ENSG00000284662   1  ENST00000332831   4    OR4F16  ensembl_havana   \n",
       "3      ENSG00000186827  11  ENST00000379236   4   TNFRSF4  ensembl_havana   \n",
       "5      ENSG00000186891  14  ENST00000328596  10  TNFRSF18  ensembl_havana   \n",
       "6      ENSG00000186891  14  ENST00000379268   7  TNFRSF18  ensembl_havana   \n",
       "8      ENSG00000160072  20  ENST00000673477   1    ATAD3B  ensembl_havana   \n",
       "...                ...  ..              ...  ..       ...             ...   \n",
       "45850  ENSG00000155313  16  ENST00000351097   9     USP25  ensembl_havana   \n",
       "45851  ENSG00000155313  16  ENST00000285681   6     USP25  ensembl_havana   \n",
       "45852  ENSG00000155313  16  ENST00000400183   7     USP25  ensembl_havana   \n",
       "45854  ENSG00000276076   5  ENST00000619537   5    CRYAA2  ensembl_havana   \n",
       "45856  ENSG00000156299  13  ENST00000286827   7     TIAM1  ensembl_havana   \n",
       "\n",
       "                   6             7               8               9      10  \\\n",
       "1      protein_coding    OR4F16-201  ensembl_havana  protein_coding   CCDS   \n",
       "3      protein_coding   TNFRSF4-201  ensembl_havana  protein_coding   CCDS   \n",
       "5      protein_coding  TNFRSF18-201  ensembl_havana  protein_coding   CCDS   \n",
       "6      protein_coding  TNFRSF18-203  ensembl_havana  protein_coding   CCDS   \n",
       "8      protein_coding    ATAD3B-206  ensembl_havana  protein_coding   CCDS   \n",
       "...               ...           ...             ...             ...    ...   \n",
       "45850  protein_coding     USP25-203  ensembl_havana  protein_coding  basic   \n",
       "45851  protein_coding     USP25-202  ensembl_havana  protein_coding   CCDS   \n",
       "45852  protein_coding     USP25-204  ensembl_havana  protein_coding   CCDS   \n",
       "45854  protein_coding    CRYAA2-201  ensembl_havana  protein_coding   CCDS   \n",
       "45856  protein_coding     TIAM1-201  ensembl_havana  protein_coding   CCDS   \n",
       "\n",
       "              11     12                                   13    14  \n",
       "1      CCDS41221  basic  NA (assigned to previous version 3)  None  \n",
       "3         CCDS11  basic   1 (assigned to previous version 3)  None  \n",
       "5          CCDS9  basic                                    1  None  \n",
       "6         CCDS10  basic   1 (assigned to previous version 6)  None  \n",
       "8         CCDS30  basic                                 None  None  \n",
       "...          ...    ...                                  ...   ...  \n",
       "45850          1   None                                 None  None  \n",
       "45851  CCDS63337  basic                                    1  None  \n",
       "45852  CCDS63336  basic   1 (assigned to previous version 6)  None  \n",
       "45854  CCDS82651  basic   1 (assigned to previous version 4)  None  \n",
       "45856  CCDS13609  basic                                    5  None  \n",
       "\n",
       "[26555 rows x 15 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(df[df['feature'] == 'transcript']['attribute']\n",
    "    .str.replace(';$', '', regex=True)\n",
    "    .str.split(';', expand=True)\n",
    "    .replace('.*\"(.*?)\"', '\\\\1', regex=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "3efb47e2-f891-4017-bb7e-2f5c05da9c83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000284662</td>\n",
       "      <td>1</td>\n",
       "      <td>OR4F16</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>11</td>\n",
       "      <td>TNFRSF4</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000186891</td>\n",
       "      <td>14</td>\n",
       "      <td>TNFRSF18</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSG00000160072</td>\n",
       "      <td>20</td>\n",
       "      <td>ATAD3B</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSG00000041988</td>\n",
       "      <td>15</td>\n",
       "      <td>THAP3</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45843</th>\n",
       "      <td>ENSG00000159200</td>\n",
       "      <td>18</td>\n",
       "      <td>RCAN1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45846</th>\n",
       "      <td>ENSG00000142197</td>\n",
       "      <td>12</td>\n",
       "      <td>DOP1B</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45848</th>\n",
       "      <td>ENSG00000155313</td>\n",
       "      <td>16</td>\n",
       "      <td>USP25</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45853</th>\n",
       "      <td>ENSG00000276076</td>\n",
       "      <td>5</td>\n",
       "      <td>CRYAA2</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45855</th>\n",
       "      <td>ENSG00000156299</td>\n",
       "      <td>13</td>\n",
       "      <td>TIAM1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>protein_coding</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19302 rows  5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     0   1         2               3               4\n",
       "0      ENSG00000284662   1    OR4F16  ensembl_havana  protein_coding\n",
       "2      ENSG00000186827  11   TNFRSF4  ensembl_havana  protein_coding\n",
       "4      ENSG00000186891  14  TNFRSF18  ensembl_havana  protein_coding\n",
       "7      ENSG00000160072  20    ATAD3B  ensembl_havana  protein_coding\n",
       "9      ENSG00000041988  15     THAP3  ensembl_havana  protein_coding\n",
       "...                ...  ..       ...             ...             ...\n",
       "45843  ENSG00000159200  18     RCAN1  ensembl_havana  protein_coding\n",
       "45846  ENSG00000142197  12     DOP1B  ensembl_havana  protein_coding\n",
       "45848  ENSG00000155313  16     USP25  ensembl_havana  protein_coding\n",
       "45853  ENSG00000276076   5    CRYAA2  ensembl_havana  protein_coding\n",
       "45855  ENSG00000156299  13     TIAM1  ensembl_havana  protein_coding\n",
       "\n",
       "[19302 rows x 5 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(df[df['feature'] == 'gene']['attribute']\n",
    "    .str.replace(';$', '', regex=True)\n",
    "    .str.split(';', expand=True)\n",
    "    .replace('.*\"(.*?)\"', '\\\\1', regex=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "cb4f920b-bd4f-4a30-b0d8-ddf2b7d5cb85",
   "metadata": {},
   "outputs": [],
   "source": [
    "del df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8dcb436c-6972-45b0-a114-c7a16944b066",
   "metadata": {},
   "source": [
    "### BED genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "0efbd065-9c9a-41a5-8bce-ca542c5c5a13",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene = df[df['feature'] == 'gene'].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "8e20943a-9510-4216-a7ad-38eba35b3578",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_id \"ENSG00000284662\"</td>\n",
       "      <td>gene_version \"1\"</td>\n",
       "      <td>gene_name \"OR4F16\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>gene_id \"ENSG00000186827\"</td>\n",
       "      <td>gene_version \"11\"</td>\n",
       "      <td>gene_name \"TNFRSF4\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>gene_id \"ENSG00000186891\"</td>\n",
       "      <td>gene_version \"14\"</td>\n",
       "      <td>gene_name \"TNFRSF18\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gene_id \"ENSG00000160072\"</td>\n",
       "      <td>gene_version \"20\"</td>\n",
       "      <td>gene_name \"ATAD3B\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>gene_id \"ENSG00000041988\"</td>\n",
       "      <td>gene_version \"15\"</td>\n",
       "      <td>gene_name \"THAP3\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19297</th>\n",
       "      <td>gene_id \"ENSG00000159200\"</td>\n",
       "      <td>gene_version \"18\"</td>\n",
       "      <td>gene_name \"RCAN1\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19298</th>\n",
       "      <td>gene_id \"ENSG00000142197\"</td>\n",
       "      <td>gene_version \"12\"</td>\n",
       "      <td>gene_name \"DOP1B\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19299</th>\n",
       "      <td>gene_id \"ENSG00000155313\"</td>\n",
       "      <td>gene_version \"16\"</td>\n",
       "      <td>gene_name \"USP25\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19300</th>\n",
       "      <td>gene_id \"ENSG00000276076\"</td>\n",
       "      <td>gene_version \"5\"</td>\n",
       "      <td>gene_name \"CRYAA2\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19301</th>\n",
       "      <td>gene_id \"ENSG00000156299\"</td>\n",
       "      <td>gene_version \"13\"</td>\n",
       "      <td>gene_name \"TIAM1\"</td>\n",
       "      <td>gene_source \"ensembl_havana\"</td>\n",
       "      <td>gene_biotype \"protein_coding\"</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19302 rows  5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               0                   1                      2  \\\n",
       "0      gene_id \"ENSG00000284662\"    gene_version \"1\"     gene_name \"OR4F16\"   \n",
       "1      gene_id \"ENSG00000186827\"   gene_version \"11\"    gene_name \"TNFRSF4\"   \n",
       "2      gene_id \"ENSG00000186891\"   gene_version \"14\"   gene_name \"TNFRSF18\"   \n",
       "3      gene_id \"ENSG00000160072\"   gene_version \"20\"     gene_name \"ATAD3B\"   \n",
       "4      gene_id \"ENSG00000041988\"   gene_version \"15\"      gene_name \"THAP3\"   \n",
       "...                          ...                 ...                    ...   \n",
       "19297  gene_id \"ENSG00000159200\"   gene_version \"18\"      gene_name \"RCAN1\"   \n",
       "19298  gene_id \"ENSG00000142197\"   gene_version \"12\"      gene_name \"DOP1B\"   \n",
       "19299  gene_id \"ENSG00000155313\"   gene_version \"16\"      gene_name \"USP25\"   \n",
       "19300  gene_id \"ENSG00000276076\"    gene_version \"5\"     gene_name \"CRYAA2\"   \n",
       "19301  gene_id \"ENSG00000156299\"   gene_version \"13\"      gene_name \"TIAM1\"   \n",
       "\n",
       "                                   3                               4  \n",
       "0       gene_source \"ensembl_havana\"   gene_biotype \"protein_coding\"  \n",
       "1       gene_source \"ensembl_havana\"   gene_biotype \"protein_coding\"  \n",
       "2       gene_source \"ensembl_havana\"   gene_biotype \"protein_coding\"  \n",
       "3       gene_source \"ensembl_havana\"   gene_biotype \"protein_coding\"  \n",
       "4       gene_source \"ensembl_havana\"   gene_biotype \"protein_coding\"  \n",
       "...                              ...                             ...  \n",
       "19297   gene_source \"ensembl_havana\"   gene_biotype \"protein_coding\"  \n",
       "19298   gene_source \"ensembl_havana\"   gene_biotype \"protein_coding\"  \n",
       "19299   gene_source \"ensembl_havana\"   gene_biotype \"protein_coding\"  \n",
       "19300   gene_source \"ensembl_havana\"   gene_biotype \"protein_coding\"  \n",
       "19301   gene_source \"ensembl_havana\"   gene_biotype \"protein_coding\"  \n",
       "\n",
       "[19302 rows x 5 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene['attribute'].str.replace(';$', '', regex=True).str.split(';', expand=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "045c7fa7-4a77-497c-899d-f34e27ed3cbd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# zcat Homo_sapiens.GRCh38.104.chr.gtf.gz | awk 'OFS=\"\\t\" {if ($3==\"gene\") {print $1,$4-1,$5,$10,$14,$7}}' | tr -d '\";' > hg38.genes.bed \n",
    "# zcat Homo_sapiens.GRCh38.104.chr.gtf.gz | awk 'OFS=\"\\t\" {if ($3==\"transcript\") {print $1,$4-1,$5,$10,$24,$7}}' | tr -d '\";' > hg38.transcripts.bed "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "75d8b654-fdcd-4f5f-b1bf-703a8b4139a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_columns = ['gene_id', 'gene_version', 'gene_name']\n",
    "df_gene[new_columns] = df_gene['attribute'].str.replace(';$', '', regex=True).str.split(';', expand=True).iloc[:, :3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "7112aae5-30a6-4f47-9299-174cf8a5c442",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_version</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>685679</td>\n",
       "      <td>686673</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000284662\"; gene_version \"1\"; g...</td>\n",
       "      <td>gene_id \"ENSG00000284662\"</td>\n",
       "      <td>gene_version \"1\"</td>\n",
       "      <td>gene_name \"OR4F16\"</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1211340</td>\n",
       "      <td>1214153</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186827\"; gene_version \"11\"; ...</td>\n",
       "      <td>gene_id \"ENSG00000186827\"</td>\n",
       "      <td>gene_version \"11\"</td>\n",
       "      <td>gene_name \"TNFRSF4\"</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1203508</td>\n",
       "      <td>1206592</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186891\"; gene_version \"14\"; ...</td>\n",
       "      <td>gene_id \"ENSG00000186891\"</td>\n",
       "      <td>gene_version \"14\"</td>\n",
       "      <td>gene_name \"TNFRSF18\"</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000160072\"; gene_version \"20\"; ...</td>\n",
       "      <td>gene_id \"ENSG00000160072\"</td>\n",
       "      <td>gene_version \"20\"</td>\n",
       "      <td>gene_name \"ATAD3B\"</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>6624866</td>\n",
       "      <td>6635586</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000041988\"; gene_version \"15\"; ...</td>\n",
       "      <td>gene_id \"ENSG00000041988\"</td>\n",
       "      <td>gene_version \"15\"</td>\n",
       "      <td>gene_name \"THAP3\"</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname          source feature    start      end score strand frame  \\\n",
       "0       1  ensembl_havana    gene   685679   686673     .      -     .   \n",
       "1       1  ensembl_havana    gene  1211340  1214153     .      -     .   \n",
       "2       1  ensembl_havana    gene  1203508  1206592     .      -     .   \n",
       "3       1  ensembl_havana    gene  1471765  1497848     .      +     .   \n",
       "4       1  ensembl_havana    gene  6624866  6635586     .      +     .   \n",
       "\n",
       "                                           attribute  \\\n",
       "0  gene_id \"ENSG00000284662\"; gene_version \"1\"; g...   \n",
       "1  gene_id \"ENSG00000186827\"; gene_version \"11\"; ...   \n",
       "2  gene_id \"ENSG00000186891\"; gene_version \"14\"; ...   \n",
       "3  gene_id \"ENSG00000160072\"; gene_version \"20\"; ...   \n",
       "4  gene_id \"ENSG00000041988\"; gene_version \"15\"; ...   \n",
       "\n",
       "                     gene_id        gene_version              gene_name  \n",
       "0  gene_id \"ENSG00000284662\"    gene_version \"1\"     gene_name \"OR4F16\"  \n",
       "1  gene_id \"ENSG00000186827\"   gene_version \"11\"    gene_name \"TNFRSF4\"  \n",
       "2  gene_id \"ENSG00000186891\"   gene_version \"14\"   gene_name \"TNFRSF18\"  \n",
       "3  gene_id \"ENSG00000160072\"   gene_version \"20\"     gene_name \"ATAD3B\"  \n",
       "4  gene_id \"ENSG00000041988\"   gene_version \"15\"      gene_name \"THAP3\"  "
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "1d143c14-9ca8-4d93-bc60-b4cdf2f20fa6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene = (df_gene[['seqname', 'start', 'end', \n",
    "                    'gene_id', 'score', 'strand', 'gene_name']]\n",
    "                 .replace('\"', \"\", regex=True)\n",
    "                 .replace(\"gene_id\", \"\", regex=True)\n",
    "                 .replace(\"gene_name\", \"\", regex=True))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "13b28911-7554-4bd8-9a6c-24fcf3eb794a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0          685679\n",
       "1         1211340\n",
       "2         1203508\n",
       "3         1471765\n",
       "4         6624866\n",
       "           ...   \n",
       "19297    34513142\n",
       "19298    36156782\n",
       "19299    15729982\n",
       "19300     6499203\n",
       "19301    31118416\n",
       "Name: start, Length: 19302, dtype: int64"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.to_numeric(df_gene['start'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "2eb8ef59-2cf1-46bf-a3cb-1f1903cee110",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     685679\n",
       "1    1211340\n",
       "2    1203508\n",
       "3    1471765\n",
       "4    6624866\n",
       "Name: start, dtype: int64"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene['start'].astype('int64').head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "ce32586e-8506-4a50-9338-e5468b1c1e83",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0          685678\n",
       "1         1211339\n",
       "2         1203507\n",
       "3         1471764\n",
       "4         6624865\n",
       "           ...   \n",
       "19297    34513141\n",
       "19298    36156781\n",
       "19299    15729981\n",
       "19300     6499202\n",
       "19301    31118415\n",
       "Name: start, Length: 19302, dtype: int64"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.to_numeric(df_gene['start'])-1 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "bea38393-a96d-4eaa-9c0a-e4840b9eed5d",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene['start'] = df_gene['start'].astype('int64') - 1 "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "1b47edd0-9dda-42d1-a4ef-eec889ec76b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>685678</td>\n",
       "      <td>686673</td>\n",
       "      <td>ENSG00000284662</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>OR4F16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1211339</td>\n",
       "      <td>1214153</td>\n",
       "      <td>ENSG00000186827</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>TNFRSF4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1203507</td>\n",
       "      <td>1206592</td>\n",
       "      <td>ENSG00000186891</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>TNFRSF18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1471764</td>\n",
       "      <td>1497848</td>\n",
       "      <td>ENSG00000160072</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>ATAD3B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>6624865</td>\n",
       "      <td>6635586</td>\n",
       "      <td>ENSG00000041988</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>THAP3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqname    start      end           gene_id score strand   gene_name\n",
       "0       1   685678   686673   ENSG00000284662     .      -      OR4F16\n",
       "1       1  1211339  1214153   ENSG00000186827     .      -     TNFRSF4\n",
       "2       1  1203507  1206592   ENSG00000186891     .      -    TNFRSF18\n",
       "3       1  1471764  1497848   ENSG00000160072     .      +      ATAD3B\n",
       "4       1  6624865  6635586   ENSG00000041988     .      +       THAP3"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "2acece4c-5c41-4153-8f75-75bbf08443a2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(19302, 7)"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16ea8df2-e7f3-46c0-aec8-cd8681cfac1a",
   "metadata": {},
   "source": [
    "#### Atributos de Pandas\n",
    "- `df.shape`\n",
    "\n",
    "#### Mtodos de pandas\n",
    "\n",
    "- `df.head()`\n",
    "- `df.tail()`\n",
    "- `df.isna()`\n",
    "- `df.sum()`\n",
    "- `df.groupby()`\n",
    "- `df.count()`\n",
    "- `df[x].unique()`\n",
    "- `df.reset_index()`\n",
    "- `df.str.contains()`\n",
    "- `df.isin()`\n",
    "- `df.replace()`\n",
    "- `df.str.replace()`\n",
    "- `df.str.split()`\n",
    "- `df.to_numeric()`\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "##### Funciones built-in\n",
    "- `len()`\n",
    "- `print()`\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
